File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if not root: return 0
        
        diameter = 0
        que = [root]
        
        def getHeight(node):
            if not node: return 0
            
            return 1 + max(getHeight(node.left), getHeight(node.right))
        
        while que:
            curNode = que.pop(0)
            left = getHeight(curNode.left)
            right = getHeight(curNode.right)
            
            if curNode.left:
                que.append(curNode.left)
            if curNode.right:
                que.append(curNode.right)
            
            diameter = max(left+right, diameter)
        
        return diameter
        
        
        #         time: o(n), space: o(1)
#         if not root: return 0
        
#         self.diameter = 0
        
#         def helper(node):
#             if not node: return 0
            
#             left = helper(node.left)
#             right = helper(node.right)
            
#             self.diameter = max(left + right, self.diameter)
            
#             return max(left, right)+1
        
#         helper(root)
#         return self.diameter

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {
    /*
    DFS
    
    1. in a helper func, if root not exist, return 0
    2. else 1 + helper func
    */
    
    if(!root) return 0;
    let max = 0;
    dfs(root);
    
    return max;
    
    function dfs(root){
        if(!root) return 0;
        let left = dfs(root.left);
        let right = dfs(root.right);
        max = Math.max(left+ right, max);
        return Math.max(right, left) + 1
    }
};